drivers: ethernet: phy: add driver for ksz9131

TonyHan11 opened the pull request:
This PR includes:
- add the driver for Microchip KSZ9131 Ethernet PHY
- add interrupt mode support for KSZ8081

TonyHan11 committed:
soc: microchip: sam: update MMU for sama7g5 GMAC0

Enable strong ordered access to the GMAC0 registers.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: ethernet: sam_gmac: update for support RGMII mode

Update to use the RGMII mode which is supported by SAMA7G54.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
soc: microchip: sam: configure the clocks for sama7g5 GMAC0

Configure the generic clocks to 125MHz for GMAC0.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: ethernet: phy: add Microchip's KSZ9131 PHY support

Add support for KSZ9131 (Gigabit Ethernet Transceiver with RGMII Support).
As first starter, 100MBit/s mode is tested.
https://www.microchip.com/en-us/product/ksz9131

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: ethernet: phy: ksz9131: add support for phy interrupt mode

Enable Link-Up and Link-Down interrupts. On the interrupt handling
the monitor work is scheduled to update the link status and calling
corresponding callback routine.

Signed-off-by: Tony Han <tony.han@microchip.com>

pdgendt commented:
Add an entry to https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/ethernet/app.overlay so it builds in CI.

TonyHan11 committed:
tests: drivers: build_all: ethernet: add KSZ9131 entry

Add build tests for Microchip KSZ9131.

Signed-off-by: Tony Han <tony.han@microchip.com>

maass-hamburg commented:
@TonyHan11 pls add PR description and address the things from sonarqube

TonyHan11 commented:
@maass-hamburg updated with removing the unused property, renaming `LINK_*BASED_T` and fixing `if (ret)`, thank you very much.

TonyHan11 commented:
@maass-hamburg
Thank you for your comments. Commits updated with the following changes:
- using `goto` instead of `do while (0)`
- update `if (ret)` to `if (ret < 0)`
- fix the issues reported by SonarQube Cloud
- remove comments `TODO change this to GPIO interrupt driven`

TonyHan11 commented on drivers/ethernet/phy/Kconfig:
```diff
@@ -51,6 +51,15 @@ config PHY_MICROCHIP_KSZ8081
+	default y
+	depends on DT_HAS_MICROCHIP_KSZ9131_ENABLED
+	depends on MDIO
+	depends on GPIO
```
Updated, thanks.

TonyHan11 commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
```diff
@@ -513,6 +614,41 @@ static int phy_mc_ksz8081_init(const struct device *dev)
 
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
+	if (!config->interrupt_gpio.port) {
+		phy_mc_ksz8081_monitor_work_handler(&data->phy_monitor_work.work);
```
Do you mean that the following needs to be added before `#endif`? Thanks.
```
#else
	phy_mc_ksz8081_monitor_work_handler(&data->phy_monitor_work.work);
```
On my side waiting network interface will timeout without `int_gpios`.
Not know why `phy_mc_ksz8081_monitor_work_handler()` is missing in current mainline.

TonyHan11 replied:
Many thanks. Updated with `return ret;` and removing `phy_mc_ksz8081_monitor_work_handler()`.

TonyHan11 commented on drivers/ethernet/phy/phy_microchip_ksz8081.c:
```diff
@@ -513,6 +614,41 @@ static int phy_mc_ksz8081_init(const struct device *dev)
+	if (ret < 0) {
+		LOG_ERR("PHY (%d) config interrupt failed", config->addr);
+	}
+#endif
 	return 0;
```
OK.

kartben commented on drivers/ethernet/phy/phy_microchip_ksz9131.c:
```diff
@@ -0,0 +1,566 @@
+	reg_val |= PHY_KSZ9131_ICS_LINK_UP_IE_MASK | PHY_KSZ9131_ICS_LINK_DOWN_IE_MASK;
+
+	/* Write settings to Interrupt Control/Status register */
+	ret = ksz9131_write(dev, 27, reg_val);
```
Use macro instead of magic number

kartben commented on drivers/ethernet/phy/phy_microchip_ksz9131.c:
```diff
@@ -0,0 +1,566 @@
+		goto done;
+	}
+
+	mutual_capabilities = anar & anlpar;
+	if (mutual_capabilities & MII_ADVERTISE_100_FULL) {
+		state->speed = LINK_FULL_100BASE;
+	} else if (mutual_capabilities & MII_ADVERTISE_100_HALF) {
+		state->speed = LINK_HALF_100BASE;
+	} else if (mutual_capabilities & MII_ADVERTISE_10_FULL) {
+		state->speed = LINK_FULL_10BASE;
+	} else if (mutual_capabilities & MII_ADVERTISE_10_HALF) {
+		state->speed = LINK_HALF_10BASE;
+	} else {
+		ret = -EIO;
+	}
```
no gigabit?

TonyHan11 committed:
drivers: ethernet: phy: ksz9131: add getting status for gigabit mode

Read gigabit status from Master Slave Status Register.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented:
@kartben Thank you for your comments.

updated with the following changes:
- replace the magic number
- add gigabit support
- remove extra `if (ret < 0) {` after `gpio_init_callback()` due to it's a `void` routine

TonyHan11 commented on drivers/ethernet/phy/phy_microchip_ksz9131.c:
```diff
@@ -0,0 +1,583 @@
+#define PHY_ID_KSZ9131				0x00221640
+#define PHY_ID_KSZ9131_MSK			(~0xF)
+
+#define PHY_KSZ9131_MSS_REG			10
+#define PHY_KSZ9131_MSS_ADVERTISE_1000_FULL	BIT(11)
+#define PHY_KSZ9131_MSS_ADVERTISE_1000_HALF	BIT(10)
```
No suitable macros found in [mii.h](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/mii.h) for these 2 bits. No sure if it's special for KSZ9131.

maass-hamburg commented on drivers/ethernet/phy/phy_microchip_ksz9131.c:
```diff
@@ -0,0 +1,583 @@
+	} else if (mssr & PHY_KSZ9131_MSS_ADVERTISE_1000_HALF) {
+		state->speed = LINK_HALF_1000BASE;
+		goto done;
+	}
```
I would prefer, if you would reuse the code of the phy_mii.c
```
	if (data->gigabit_supported) {
		if (phy_mii_reg_read(dev, MII_1KTCR, &c1kt_reg) < 0) {
			return -EIO;
		}
		if (phy_mii_reg_read(dev, MII_1KSTSR, &s1kt_reg) < 0) {
			return -EIO;
		}
		s1kt_reg = (uint16_t)(s1kt_reg >> MII_1KSTSR_OFFSET);
	}

	if (data->gigabit_supported &&
			((c1kt_reg & s1kt_reg & MII_ADVERTISE_1000_FULL) != 0U)) {
		data->state.speed = LINK_FULL_1000BASE;
	} else if (data->gigabit_supported &&
			((c1kt_reg & s1kt_reg & MII_ADVERTISE_1000_HALF) != 0U)) {
		data->state.speed = LINK_HALF_1000BASE;
```
and just shift the `MII_1KSTSR` register by 2.

Btw good way to only read the 100/10 Mbit/s regs, when gigabit doesn't match.

TonyHan11 commented:
Rebased to the latest `main` branch.

maass-hamburg commented on drivers/ethernet/phy/phy_microchip_ksz9131.c:
```diff
@@ -0,0 +1,655 @@
+	struct k_work_delayable monitor_work;
+	struct phy_link_state state;
+	struct k_sem sem;
+	bool gigabit_supported;
```
gigabit_supported bool is not needed, as this phy definetly supports gigabit and it be always be true, it is only in the generic phy driver, because there we don't know it because we don't know durring build if gigabit is supported.

TonyHan11 replied:
Updated with removing `gigabit_supported` and code related. Thanks.

maass-hamburg commented on drivers/ethernet/phy/phy_microchip_ksz9131.c:
```diff
@@ -0,0 +1,622 @@
+		LOG_ERR("PHY (%d) init failed", cfg->phy_addr);
+	}
+
+	return ret;
```
apply changes from https://github.com/zephyrproject-rtos/zephyr/pull/91572 also to this phy

TonyHan11 replied:
Updated with adding `default-speeds` for ksz9131, thanks.

TonyHan11 commented on drivers/ethernet/phy/phy_microchip_ksz9131.c:
```diff
@@ -0,0 +1,628 @@
+		goto done;
+	}
+
+	ret = phy_mchp_ksz9131_autonegotiate(dev);
```
Improve some of the logging.
No changes with auto-negotiate due to no block issue in ksz9131 using interrupt mode.
It seems that the interrupt mode for ksz8081 is not fully supported in the PR.

TonyHan11 commented:
> Dropped the commit for adding interrupt mode support for ksz8081 as it has conflicts with the code on mainline. Will add it in a new PR once it works again.

Support for ksz8081 interrupt mode added in https://github.com/zephyrproject-rtos/zephyr/pull/95305

TonyHan11 commented:
Updated ksz9131 driver with:
 - use `if (USING_INTERRUPT_GPIO)` to make code more concise
 - use `__maybe_unused` for variables in stead of using `#if .. #endif`
 - split init_int_gpios from `phy_mchp_ksz9131_init()`

maass-hamburg commented on drivers/ethernet/phy/phy_microchip_ksz9131.c:
```diff
@@ -0,0 +1,658 @@
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	ret = phy_mchp_ksz9131_update_anar(dev, adv_speeds);
+	if (ret < 0) {
+		goto done;
+	}
+
+	ret = ksz9131_read(dev, MII_1KTCR, &c1kt);
+	if (ret < 0) {
+		goto done;
+	}
+
+	if (adv_speeds & LINK_FULL_1000BASE) {
+		c1kt |= MII_ADVERTISE_1000_FULL;
+	} else {
+		c1kt &= ~MII_ADVERTISE_1000_FULL;
+	}
+
+	if (adv_speeds & LINK_HALF_1000BASE) {
+		c1kt |= MII_ADVERTISE_1000_HALF;
+	} else {
+		c1kt &= ~MII_ADVERTISE_1000_HALF;
+	}
+
+	ret = ksz9131_write(dev, MII_1KTCR, c1kt);
+	if (ret < 0) {
+		goto done;
+	}
```
if you don't want to use `phy_mii_cfg_link_autoneg`, at least use `phy_mii_set_anar_reg` and `phy_mii_set_c1kt_reg`

TonyHan11 replied:
Replaced this lines with `phy_mii_set_anar_reg` and `phy_mii_set_c1kt_reg`, thanks a lot.

TonyHan11 committed:
drivers: ethernet: phy: ksz9131: save link state for get

Get the link state in the monitor and save it for get_link api
implementation to use.

Signed-off-by: Tony Han <tony.han@microchip.com>

nandojve commented:
Hi @TonyHan11 ,

Is this still relevant to v4.3 ?

TonyHan11 committed:
dts: microchip: sam: add nodes for GMAC0 to sama7g5.dtsi

Add gmac1 and gmac1_mdio nodes.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
boards: microchip: sam: enable nodes for GMAC0 to sama7g54_ek dts file

Enable gmac0, gmac0_mdio and gmac0_phy nodes.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
tests: drivers: build_all: ethernet: add new lines between PHY nodes

New lines are expected between PHY nodes by DevicetreeLinting.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
boards: microchip: sam: remove extra new lines from sama7g54_ek.dts

Remove extra new lines to satisfy the requirement of DevicetreeLinting.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented:
Hi @nandojve ,
> Hi @TonyHan11 ,
> 
> Is this still relevant to v4.3 ?

Yes, we hope it be merged to v4.3.
I rebased the commits to totay's `main` branch with 2 errors raised by `DevicetreeLinting` fixed. Also add some commits on it for supporting the `GMAC0` which uses KSZ9131 as the PHY.

nandojve reviewed (dismissed):
I took a general look and LGTM!

TonyHan11 commented on drivers/ethernet/phy/Kconfig:
```diff
@@ -56,6 +56,16 @@ config PHY_MICROCHIP_KSZ8081
+	depends on DT_HAS_MICROCHIP_KSZ9131_ENABLED
+	select MDIO
+	select GPIO if ($(dt_compat_any_has_prop,$(DT_COMPAT_MICROCHIP_KSZ9131),reset-gpios) || \
+			$(dt_compat_any_has_prop,$(DT_COMPAT_MICROCHIP_KSZ9131),int-gpios))
```
`reset-gpios` removed from Kconfig, thanks.
